### PR TITLE
Force embind global initializer at the end of __ATINIT__

### PR DESCRIPTION
**Pull request by AlexPerrot**
This PR implements the second solution I proposed in https://github.com/kripken/emscripten/issues/3436.
This works well on my code and tests are passing.

I would have liked to be able to check if `--bind` was passed, but couldn't find how...

### CODE REVIEW

File: emscripten.py [hunk 1]
```diff
@@ -320,7 +320,16 @@ def function_tables_and_exports(funcs, metadata, mem_init, glue, forwarded_data,
 
     # memory and global initializers
 
-    global_initializers = str(', '.join(map(lambda i: '{ func: function() { %s() } }' % i, metadata['initializers'])))
+    initializers_list = metadata['initializers']
+
+    # make sure embind's global initializer is at the end of the list
+    # fixes problems with dependent types resolution order
+    # see https://github.com/kripken/emscripten/issues/3436
+    embind_global_initializer_name = "__GLOBAL__sub_I_bind_cpp"
```
Comment (kripken) [inline emscripten.py]: single quotes please

**Comment by kripken** [comment]
cc @chadaustin @waywardmonkeys 

This worries me, it depends on the specific name - which we have no guarantee of staying the same, it's a clang internal detail. Hopefully there is another solution to this issue?

**Comment by AlexPerrot** [comment]
PR #3450 has another solution for this, which simply fixes embind to be independent of the static initialization order.
I Hope we can find a reasonable way to fix this.
For now, i'm using `sed` to post process the js and change the generated `__ATINIT__`.

**Comment by juj** [comment]
Some great debugging in getting to the root cause of this bug. I can see this has been a painful one.

The execution order of static initializers is undefined between compilation units in all C++ standards. This fix is attempting to define and ordering between the compilation units, which I don't think is good. I think the correct fix should be to fix the initialization ordering at C++ level. [StackOverflow](http://stackoverflow.com/questions/211237/static-variables-initialisation-order) has a good description.

The PR #3450 looks like it does not fix the initialization order, but attempts to make the code resilient to running no matter which initialization order is taken? If so, that sounds a bit fishy, since it could be accessing yet uninitialized data anyways, even though it seems to work. (not completely sure about this though)

Ideally, I'd hope making use of a [singleton accessor](http://stackoverflow.com/a/1463787) to model the initialization order dependency explicitly would be the cleanest fix. Is that something that might work here?

If that doesn't cut it for some reason, then perhaps LLVM supports the same `init_priority` attribute that GCC does (https://gcc.gnu.org/onlinedocs/gcc/C_002b_002b-Attributes.html)? That could also be used to enforce the initialization order. If I read this correctly, that should do the exact same thing as this PR, except that it'd solve the issue at C++ level instead of needing to do specialities in `.py` side.

Does the test you added in https://github.com/kripken/emscripten/pull/3450/files expose the bug in the suite?

**Comment by AlexPerrot** [comment]
This [post](https://github.com/kripken/emscripten/issues/3436#issuecomment-98998708) has a pretty good description of what is happening.
I made a sequence diagram to show the order of operations
embind overload registration

I don't think it is a bug that the `unboundTypesHandler` is replaced right away in the first call. But the `argsCount` should be kept in case an overload is registered later. This doesn't really need anything at the C++ level IMO.

@juj 

> The PR #3450 looks like it does not fix the initialization order, but attempts to make the code resilient to running no matter which initialization order is taken? If so, that sounds a bit fishy, since it could be accessing yet uninitialized data anyways, even though it seems to work. (not completely sure about this though)

Well, since static initialization order is undefined, shouldn't the code be resilient to any order ;-) ? IIRC, this part of embind only accesses data it has created or verified existed, so I'm not concerned about this.

> Does the test you added in https://github.com/kripken/emscripten/pull/3450/files expose the bug in the suite?

It has been a while, but reading the code, I think it does. However, It uses user defined types only, so this PR doesn't solve the problem highlighted in the test. Testing this with builtin types is tricky since LLVM needs to output the static initializers in a specific way in order for the bug to show up (which this PR does fix).

**Comment by juj** [comment]
> Well, since static initialization order is undefined, shouldn't the code be resilient to any order ;-) ?

Yeah, that is why I recommended using a singleton approach to enforcing the initialization order, or using the weak initializer attribute to define the order. Expanding `emscripten.py` to define its own initializer orders doesn't sound like a good idea, since it would be an ad hoc machinery to get over this issue.

**Comment by juj** [comment]
Hmm, reading further and referencing the diagram (excellent drawing btw), it looks to me that to PR #3450 would be an appropriate fix.

**Comment by juj** [comment]
The way that `EMSCRIPTEN_BINDINGS()` works is it runs in unspecified order for registering all types, so the JS side should be prepared to handle any ordering of the registration. The PR #3450 seems to do fix that correctly I think.

**Comment by AlexPerrot** [comment]
Closing this, since #3450 has been merged.

**Pull request closed without merging**